docs: programs.md (#2908)
anza-xyz/agave: docs/src/runtime/programs.md

---

Solana contains a small handful of native programs, which are required to run

> validator nodes. Unlike the third-party programs, the native programs are part of

Reviewer: I don't think the grammar was wrong here. In fact, this actually makes less sense because the "the" makes the reader think there are special third-party programs being referred to here.

This is almost pedantic, but I think maybe drop this change please.

## Config Program

> Add configuration data to the chain, alongside the list of public keys that are allowed to modify it

Reviewer: (lines not wrapped to 80 in my below suggestion)
```suggestion
Add configuration data to the chain, followed by the list of public keys that are allowed to modify it
```

- Program id: `Config1111111111111111111111111111111111111`
- Instructions: [config_instruction](https://docs.rs/solana-config-program/VERSION_FOR_DOCS_RS/solana_config_program/config_instruction/index.html)

Unlike the other programs, the Config program does not define any individual
instructions. It has just one implicit instruction: "store". Its
instruction data is a set of keys that gate access to the account and the

> data stored inside it.

Reviewer: ```suggestion
data stored inside of it.
```

Reviewer 2: The instruction data is a set of keys and the data to store in the account. The change makes it seem like the instruction data is the data already stored in the account, which is incorrect. Maybe "the data to store inside of it" is clearest?

## Config Program

> Add configuration data to the chain, followed by the list of public keys that 

Reviewer 2: Need to remove trailing whitespace here
```suggestion
Add configuration data to the chain, followed by the list of public keys that
```

are allowed to modify it

- Program id: `Config1111111111111111111111111111111111111`
- Instructions: [config_instruction](https://docs.rs/solana-config-program/VERSION_FOR_DOCS_RS/solana_config_program/config_instruction/index.html)

Unlike the other programs, the Config program does not define any individual
instructions. It has just one implicit instruction: "store". Its
instruction data is a set of keys that gate access to the account and the

> data to stored inside of it.

Reviewer 2: Little typo
```suggestion
data to store inside of it.
```

- Program id: `BPFLoaderUpgradeab1e11111111111111111111111`
- Instructions: [LoaderInstruction](https://docs.rs/solana-sdk/VERSION_FOR_DOCS_RS/solana_sdk/loader_upgradeable_instruction/enum.UpgradeableLoaderInstruction.html)

The BPF Upgradeable Loader marks itself as the "owner" of the executable and
program-data accounts it creates to store your program. When a user invokes an
instruction via a program id, the Solana runtime loads both your program
and its owner, the BPF Upgradeable Loader. The runtime then passes your program
to the BPF Upgradeable Loader for it to process the instruction.

[More information about deployment](../cli/examples/deploy-a-program.md)

## Ed25519 Program

The program for verifying ed25519 signatures. It takes an ed25519 signature, a public key, and a message.
Multiple signatures can be verified. If any of the signatures fail to verify, an error is returned.

- Program id: `Ed25519SigVerify111111111111111111111111111`

> - Instructions: [ed25519_instruction](https://docs.rs/solana-sdk/1.18.18/solana_sdk/ed25519_instruction/index.html)

Reviewer 2: To properly link to the docs.rs page, let's use the dynamic version rather than a static link
```suggestion
- Instructions: [ed25519_instruction](https://docs.rs/solana-sdk/VERSION_FOR_DOCS_RS/solana_sdk/ed25519_instruction/index.html)
```

}
```

> The pseudo code of the verification:

Reviewer: ```suggestion
The pseudo code of the recovery verification:
```

### Optimization notes

> The operation will have to take place after (at least, partial) deserialization,

Reviewer: This comma doesn't belong here. The original is correct.

but all inputs come from the transaction data itself, which allows it to be

> relatively easy to execute in parallel to the transaction processing and the PoH

Reviewer: Same here, I don't think these "the" additions make sense tbh.

### Optimization notes

The operation will have to take place after (at least partial) deserialization,
but all inputs come from the transaction data itself, which allows it to be

> relatively easy to execute in parallel to the transaction processing and the PoH

Reviewer 2: Looks like this one got missed
```suggestion
relatively easy to execute in parallel to transaction processing and PoH
```
